### PR TITLE
Feat/private types

### PR DESCRIPTION
**Pull request by ipatka**
- Remove type IDs
- Integrate BloomID v2 merkle tree with enhanced user privacy
- Test all hashingLogic functions

### CODE REVIEW

File: src/HashingLogic.ts [hunk 1]
```diff
@@ -1,22 +1,22 @@
 import {AttestationTypeID} from './AttestationTypes'
 import {sortBy} from 'lodash'
 import {keccak256} from 'js-sha3'
+var crypto = require('crypto')
```
Comment (edhedges) [inline src/HashingLogic.ts]: `var` to `const`

File: src/HashingLogic.ts [hunk 2]
```diff
@@ -53,35 +120,198 @@ export const orderedStringify = (obj: {}) => {
...
 /**
- * Given an array of IAttestationData, creates a new MerkleTree with the attestations
- * after the leaves are sorted by hash and mapped into hash Buffers.
+ *
+ * @param attestation Given the contents of an attestation node, return a
+ * Merkle tree
  */
-export const getMerkleTree = (attestations: IAttestationData[]) => {
-  const leaves = attestations.map(hashAttestation)
+export const getDataTree = (attestation: IAttestationNode): MerkleTree => {
```
Comment (edhedges) [inline src/HashingLogic.ts]: Why the name change?

Comment (edhedges) [reply src/HashingLogic.ts]: Nevermind, I hadn't gotten down far enough.

File: src/HashingLogic.ts [hunk 3]
```diff
@@ -53,35 +120,198 @@ export const orderedStringify = (obj: {}) => {
...
+  dataNode: IAttestation,
+  globalRevocationLink: string,
+  privKey: Buffer
+): IDataNode => {
+  const attestationNode: IAttestationNode = {
+    data: dataNode.data,
+    type: dataNode.type,
+    aux: dataNode.aux,
+    link: {
+      local: generateNonce(),
```
Comment (edhedges) [inline src/HashingLogic.ts]: Will this `local` link nonce need to be stored locally / in our db or anything?

File: src/HashingLogic.ts [hunk 4]
```diff
@@ -53,35 +120,198 @@ export const orderedStringify = (obj: {}) => {
...
+    attestationNode: attestationNode,
+    signedAttestation: attestationSig,
+  }
+}
+
+/**
+ * Given the number of data nodes return an array of padding nodes
+ * @param {number} dataCount - number of data nodes in tree
+ */
+export const getPadding = (dataCount: number): string[] => {
```
Comment (edhedges) [inline src/HashingLogic.ts]: This function is not super easy to read. I suggest, adding some explanation docs, possibly a table explaining the first 2-3 increments of the while loop.

File: src/HashingLogic.ts [hunk 5]
```diff
@@ -53,35 +120,198 @@ export const orderedStringify = (obj: {}) => {
...
+ * Given attestation data and the attester's private key, construct the entire Bloom Merkle tree
+ * and return the components needed to generate proofs
+ * @param dataNodes - Complete attestation nodes
+ * @param privKey - Attester private key
+ */
+export const getSignedMerkleTreeComponents = (
+  dataNodes: IAttestation[],
+  privKey: Buffer
+): IBloomMerkleTreeComponents => {
+  const globalRevocationLink = generateNonce()
```
Comment (edhedges) [inline src/HashingLogic.ts]: Will this need to be stored locally / in our db?

File: src/HashingLogic.ts [hunk 6]
```diff
@@ -126,6 +356,8 @@ export const verifyMerkleProof = (
 /**
  * Given an array of type `AttestationTypeID`, sorts the array, and
  * uses `soliditySha3` to hash the array.
+ *
+ * Deprecated due to types being removed from Attestation Logic contract
  */
 export const hashAttestationTypes = (types: AttestationTypeID[]) =>
```
Comment (edhedges) [inline src/HashingLogic.ts]: We should remove if deprecated IMO

File: src/HashingLogic.ts [hunk 7]
```diff
@@ -204,9 +427,7 @@ export enum ChainId {
 }
 
 export const getAttestationAgreementEIP712 = (
```
Comment (edhedges) [inline src/HashingLogic.ts]: Did the EIP change? Why remove the `chainId` and `verifyingContract` params?

File: src/ethereumjs-wallet.d.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+declare module 'ethereumjs-wallet' {
```
Comment (edhedges) [inline src/ethereumjs-wallet.d.ts]: If this covers the whole module, do you plan on adding it to their repo at some point?

File: src/HashingLogic.ts (hunk 1 above)
Comment (ipatka) [reply src/HashingLogic.ts]: 👍

File: src/HashingLogic.ts (hunk 3 above)
Comment (ipatka) [reply src/HashingLogic.ts]: Everything returned by the getComponents function needs to be stored

File: src/HashingLogic.ts (hunk 4 above)
Comment (ipatka) [reply src/HashingLogic.ts]: Sounds good. I'll add comments

File: src/HashingLogic.ts (hunk 5 above)
Comment (ipatka) [reply src/HashingLogic.ts]: yes we should store the whole thing returned by getComponents as a JSONB

File: src/HashingLogic.ts (hunk 6 above)
Comment (ipatka) [reply src/HashingLogic.ts]: ok will do

File: src/HashingLogic.ts (hunk 7 above)
Comment (ipatka) [reply src/HashingLogic.ts]: They're optional and I didn't think they were super relevant for us

File: src/ethereumjs-wallet.d.ts (hunk 8 above)
Comment (ipatka) [reply src/ethereumjs-wallet.d.ts]: Not sure if it's full coverage but good idea

File: src/HashingLogic.test.ts [hunk 9]
```diff
@@ -129,60 +181,275 @@ test('HashingLogic orderedStringify', () => {
   )
 })
 
-test('HashingLogic merkle trees / proofs', () => {
-  const fullNameAttestation: HashingLogic.IAttestationData = {
-    type: 'full-name',
```
Comment (edhedges) [inline src/HashingLogic.test.ts]: RIP von Mises :(

File: src/HashingLogic.test.ts [hunk 10]
```diff
@@ -129,60 +181,275 @@ test('HashingLogic orderedStringify', () => {
...
+  // If 10 data node padding should be 5
+  padding = HashingLogic.getPadding(10)
+  padding.forEach(p => expect(p.length).toBe(66))
+  expect(padding.length).toBe(5)
+
+  // If 15 data node padding should be 0
+  padding = HashingLogic.getPadding(15)
+  expect(padding.length).toBe(0)
+
+  // If 16 data node padding should be 495
```
Comment (edhedges) [inline src/HashingLogic.test.ts]: this is crazy, but do we actually foresee this happening in practice?

File: src/HashingLogic.test.ts [hunk 11]
```diff
@@ -129,60 +181,275 @@ test('HashingLogic orderedStringify', () => {
...
+  padding = HashingLogic.getPadding(16)
+  padding.forEach(p => expect(p.length).toBe(66))
+  expect(padding.length).toBe(495)
+
+  // If 511 data node padding should be 0
+  padding = HashingLogic.getPadding(511)
+  expect(padding.length).toBe(0)
+
+  // If 512 data node padding should be 16383
```
Comment (edhedges) [inline src/HashingLogic.test.ts]: whoa

File: src/HashingLogic.test.ts (hunk 10 above)
Comment (ipatka) [reply src/HashingLogic.test.ts]: Perhaps when we have super-meta-corporate attestations? might as well be prepared. probably won't see more than a few data nodes for a while

**Comment by edhedges** [comment]
@ipatka One other thing you'll want to do is bump the version and publish to npm. Let me or Jace know if you need any assistance with versioning / npm. Since this is breaking we'll want to bump the major version so 2.0.0.

**Comment by ipatka** [comment]
ok great I'll update to 2.0.0. I think I know how to do npm publish but I'll let you know

File: src/HashingLogic.ts [hunk 12]
```diff
@@ -40,6 +40,73 @@ export interface IAttestationData {
...
+  nonce: string
+}
+
+export interface IAttestation {
+  data: IAttestationData
+  type: IAttestationType
+  /**
+   * aux either contains a hash of IAuxSig or just a padding node hash
+   */
+  aux: string
```
Comment (djvs) [inline src/HashingLogic.ts]: Maybe this should be optional?  Padding hash could give confusing result re: is there a signature that someone is unwilling to provide vs. is there no signature

File: src/HashingLogic.ts [hunk 13]
```diff
@@ -53,35 +120,210 @@ export const orderedStringify = (obj: {}) => {
...
+export const getPadding = (dataCount: number): string[] => {
+  if (dataCount < 1) return []
+  let i = 5
+  while (dataCount + 1 > 2 ** (i - 1)) {
+    i += 5
+  }
+  const paddingCount = 2 ** (i - 1) - (dataCount + 1)
+  return Array.apply(null, Array(paddingCount)).map(
+    (item: number, index: number) => {
+      return hashMessage(crypto.randomBytes(20))
```
Comment (djvs) [inline src/HashingLogic.ts]: output of crypto.randombytes should be stored so the attester knows he's not being spoofed

File: src/HashingLogic.ts (hunk 12 above)
Comment (ipatka) [reply src/HashingLogic.ts]: I think we want it to be ambiguous

File: src/HashingLogic.ts (hunk 13 above)
Comment (ipatka) [reply src/HashingLogic.ts]: spoofed meaning that a padding node could be a disguised data node? I think that's handled by using the checksum node and the fact that each node is signed

File: src/HashingLogic.ts (hunk 12 above)
Comment (ipatka) [reply src/HashingLogic.ts]: although I supposed it's possible to hide an alt data node as this aux node. to fix this I added dataHash and typeHash to the link node as a local checksum

**Pull request merged**